go/parquet/metadata/statistics.go

```diff
@@ -0,0 +1,517 @@
+		return bytes.Compare(a, b) == -1
+	}
+
+	sa := *(*[]int8)(unsafe.Pointer(&a))
+	sb := *(*[]int8)(unsafe.Pointer(&b))
+	i := 0
+	for ; i < len(sa) && i < len(sb); i++ {
+		if sa[i] < sb[i] {
+			return true
+		}
+		if sb[i] < sa[i] {
+			return false
+		}
+	}
+	return i == len(sa) && i != len(sb)
+}
+
+func (BooleanStatistics) cleanStat(minMax minmaxPairBoolean) *minmaxPairBoolean { return &minMax }
+func (Int32Statistics) cleanStat(minMax minmaxPairInt32) *minmaxPairInt32       { return &minMax }
+func (Int64Statistics) cleanStat(minMax minmaxPairInt64) *minmaxPairInt64       { return &minMax }
+func (Int96Statistics) cleanStat(minMax minmaxPairInt96) *minmaxPairInt96       { return &minMax }
+
+// in the case of floating point types, the following rules are applied as per parquet-mr:
+// - if any of min/max is NaN, return nothing
+// - if min is 0.0f replace with -0.0f
+// - if max is -0.0f replace with 0.0f
+func (Float32Statistics) cleanStat(minMax minmaxPairFloat32) *minmaxPairFloat32 {
+	if math.IsNaN(float64(minMax[0])) || math.IsNaN(float64(minMax[1])) {
+		return nil
+	}
+
+	if minMax[0] == math.MaxFloat32 && minMax[1] == -math.MaxFloat32 {
+		return nil
+	}
+
+	var zero float32 = 0
+	if minMax[0] == zero && !math.Signbit(float64(minMax[0])) {
+		minMax[0] = -minMax[0]
+	}
+
+	if minMax[1] == zero && math.Signbit(float64(minMax[1])) {
+		minMax[1] = -minMax[1]
+	}
+
+	return &minMax
+}
+
+func (Float64Statistics) cleanStat(minMax minmaxPairFloat64) *minmaxPairFloat64 {
+	if math.IsNaN(minMax[0]) || math.IsNaN(minMax[1]) {
+		return nil
+	}
+
+	if minMax[0] == math.MaxFloat64 && minMax[1] == -math.MaxFloat64 {
+		return nil
+	}
+
+	var zero float64 = 0
+	if minMax[0] == zero && !math.Signbit(minMax[0]) {
+		minMax[0] = -minMax[0]
+	}
+
+	if minMax[1] == zero && math.Signbit(minMax[1]) {
+		minMax[1] = -minMax[1]
+	}
+
+	return &minMax
+}
+
+func (ByteArrayStatistics) cleanStat(minMax minmaxPairByteArray) *minmaxPairByteArray {
+	if minMax[0] == nil || minMax[1] == nil {
+		return nil
+	}
+	return &minMax
+}
+
+func (FixedLenByteArrayStatistics) cleanStat(minMax minmaxPairFixedLenByteArray) *minmaxPairFixedLenByteArray {
+	if minMax[0] == nil || minMax[1] == nil {
+		return nil
+	}
+	return &minMax
+}
+
+func GetStatValue(typ parquet.Type, val []byte) interface{} {
+	switch typ {
+	case parquet.Types.Boolean:
+		return val[0] != 0
+	case parquet.Types.Int32:
+		return int32(binary.LittleEndian.Uint32(val))
+	case parquet.Types.Int64:
+		return int64(binary.LittleEndian.Uint64(val))
+	case parquet.Types.Int96:
+		p := parquet.Int96{}
+		copy(p[:], val)
+		return p
+	case parquet.Types.Float:
+		return math.Float32frombits(binary.LittleEndian.Uint32(val))
+	case parquet.Types.Double:
+		return math.Float64frombits(binary.LittleEndian.Uint64(val))
+	case parquet.Types.ByteArray:
+		fallthrough
+	case parquet.Types.FixedLenByteArray:
+		return val
+	}
+	return nil
+}
+
+func (s *ByteArrayStatistics) UpdateWithArrow(values array.Interface) {
```

Comment:
oh no, this leaked :(  

Not there are a couple of bugs some fixed some not:
1.  s.nvalues for list types isn't quite correct repeated types as it needs to account for empty parent lists.
2. dictionary arrays up until recently where not handled correctly.

Reply:
Does it make sense to move this up to where-ever arrow is being written and modify APIs so they can take a final calculated statistics (if not that is OK as we have this cruft in C++).

Reply:
hmm. the difficulty with that is that the column writer handles the paging in the implementation so that the part that writes arrow doesn't have to deal with that level of handling (it just passes the array of data to the writer and the writer handles splitting things into data pages). so it might be kinda difficult to work with a "final calculated" statistics. 

But what i'll do is I'll leave this out for now and then when i make the PR for the arrow writing portion, i'll revisit this at that point and see if i can use this suggestion or if i just have to add it back. That work for you?

Reply:
Yeah, not a blocker.